fix: many things ...

OlivierBBB opened the pull request:
WIP

OlivierBBB committed:
fix: many things ... that then broke after fetching :/

- CreateSection now "deploys" from the pov of DeploymentInfo
- stuff in the RomLex module around CREATEs
- operations on accountSnapshots now directly modify everything
- TxInitSection

OlivierBBB committed:
fix: fewer failing tests + some ROM_LEX stuff

OlivierBBB committed:
fix(selfdestruct): some fixes

OlivierBBB committed:
spotless

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -655,11 +658,11 @@ public void traceContextReEnter(MessageFrame frame) {
   public void traceContextExit(MessageFrame frame) {
     this.currentFrame().initializeFrame(frame); // TODO: is it needed ?
+
+    exitDeploymentFromDeploymentInfoPointOfView();
```
method extraction + the update is done only when necessary

OlivierBBB replied:
**This method seems to be failing on some tests ...**

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -655,11 +658,11 @@ public void traceContextReEnter(MessageFrame frame) {
-      this.transients
-          .conflation()
-          .deploymentInfo()
-          .unmarkDeploying(this.currentFrame().byteCodeAddress());
```
This needed to happen at every depth + only when necessary

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -669,6 +672,8 @@ public void traceContextExit(MessageFrame frame) {
       }
 
+      defers.resolveUponExitingContext(this, currentFrame());
+
```
We were somehow missing this ?!

letypequividelespoubelles replied:
no, it's already done at the end of `traceContextExit` 
![image](https://github.com/user-attachments/assets/8dbb8ebe-50ad-466a-bac7-b90596d0af72)

letypequividelespoubelles replied:
we need to do it for any depth, I delete your add.

letypequividelespoubelles replied:
https://github.com/Consensys/linea-tracer/pull/1027/commits/b859713fabb537eefc15cb75e16ade62fa630f0b

OlivierBBB replied:
Indeed

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -705,6 +710,25 @@ public void traceContextExit(MessageFrame frame) {
+          .markAsNotUnderDeployment(this.currentFrame().byteCodeAddress());
+    }
+  }
+
```
Extraction of method + sanity check

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -1072,4 +1097,20 @@ public void squashParentFrameReturnData() {
+  public final boolean currentDeploymentStatus() {
+    return deploymentStatusOf(this.messageFrame().getContractAddress());
+  }
+
```
Quality of life

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -1072,4 +1097,20 @@ public void squashParentFrameReturnData() {
+
+  public final boolean deploymentStatusOf(Address address) {
+    return transients.conflation().deploymentInfo().getDeploymentStatus(address);
+  }
```
Quality of life

OlivierBBB committed:
fix(StopSection): was providing wrong address in sanity check

OlivierBBB committed:
fix: miner -> coinbase

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -48,53 +49,69 @@ public TxInitializationSection(Hub hub, WorldView world) {
+              deploymentInfo.getDeploymentStatus(recipientAddress));
+    }
+
+    if (isDeployment) {
```
doing the deployment from the `DeploymentInfo` pov

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -48,53 +49,69 @@ public TxInitializationSection(Hub hub, WorldView world) {
+
+    if (isDeployment) {
+      deploymentInfo.newDeploymentAt(recipientAddress);
     }
```
doing the deployment from the DeploymentInfo pov

OlivierBBB committed:
ras

OlivierBBB committed:
ras

OlivierBBB committed:
ras

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -301,4 +327,21 @@ public void commit(List<MappedByteBuffer> buffers) {
+
+    final RomChunk chunk = new RomChunk(contractMetadata, false, true, byteCode);
+    this.chunks.add(chunk);
+  }
```
The deployment info has been updated when appropriate in `traceContextExit`.

OlivierBBB committed:
fix(romlex): fixed address of EXTCODECOPY

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -210,29 +210,29 @@ public void callRomLex(final MessageFrame frame) {
+        checkArgument(!foreignDeploymentStatus, "EXTCODECOPY should only trigger a ROM_LEX chunk if its target isn't currently deploying");
+        checkArgument(!frame.getWorldUpdater().getAccount(foreignCodeAddress).isEmpty() && frame.getWorldUpdater().getAccount(foreignCodeAddress).hasCode());
+
+        Optional.ofNullable(frame.getWorldUpdater().get(foreignCodeAddress))
```
EXTCODECOPY foreign address fix: `frame.getContractAddress()` returns the address of the currently executing bytecode. For the `RomChunk` we are building for `EXTCODECOPY` we only care about the `foreignAddress`.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -241,15 +241,22 @@ public void callRomLex(final MessageFrame frame) {
-    this.createDefers.trigger(contractMetadata);
+    final RomChunk chunk = new RomChunk(contractMetadata, true, false, byteCode);
+    chunks.add(chunk);
+    createDefers.trigger(contractMetadata);
```
CREATE's are deferred to `ContextEntry`

I kept 
```java
createDefers.trigger(contractMetadata);
```
as I didn't look into what it does.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -46,7 +51,8 @@
 @Accessors(fluent = true)
-public class RomLex implements Module, PostOpcodeDefer {
+public class RomLex
+    implements Module, PostOpcodeDefer, ImmediateContextEntryDefer, ContextExitDefer {
```
`ImmediateContextEntryDefer` for CREATE(2)
`ContextExitDefer` for RETURN

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/transients/DeploymentInfo.java:
```diff
@@ -32,28 +36,45 @@ public class DeploymentInfo {
+  public void newDeploymentAtForTxSkip(Address address) {
+    this.incrementDeploymentNumber(address);
+    this.markAsNotUnderDeployment(address);
+  }
```
`newDeploymentAtForTxSkip` is special as one immediately transitions to a deployed account with empty code.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/SelfdestructSection.java:
```diff
@@ -69,27 +70,27 @@ public SelfdestructSection(Hub hub) {
-    this.address = frame.getSenderAddress();
-    this.accountBefore = AccountSnapshot.canonical(hub, this.address);
+    address = frame.getRecipientAddress();
+    selfdestructorAccountBefore = AccountSnapshot.canonical(hub, address);
```
Fixed major mix up where the address to  SELFDESTRUCT was declared to be the caller (or creator) rather than the callee (or createe) 😅

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/SelfdestructSection.java:
```diff
@@ -69,27 +70,27 @@ public SelfdestructSection(Hub hub) {
-            ? this.accountAfter.deepCopy()
-            : AccountSnapshot.canonical(hub, this.recipientAddress);
+            ? selfdestructorAccountAfter.deepCopy()
+            : AccountSnapshot.canonical(hub, recipientAddress);
```
**Question:** will this work for the deployment number and status ? We must have tests with SELFDESTRUCT's during CREATE(2)'s or during deployment transactions.

OlivierBBB committed:
ras

letypequividelespoubelles committed:
delete duplicate triggering exit defers + no need to trigger contextExit of module, never implemented

Signed-off-by: Francois Bojarski <francois.bojarski@consensys.net>

OlivierBBB commented:
Connected to both

https://github.com/Consensys/linea-tracer/issues/1004
https://github.com/Consensys/linea-tracer/issues/1006

OlivierBBB committed:
fix(EXTCODECOPY): trigger ROM_LEX only when required

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 1006-range-6032696-6032699

letypequividelespoubelles committed:
spotless + renable failing tests

Signed-off-by: Francois Bojarski <francois.bojarski@consensys.net>

letypequividelespoubelles committed:
test: disabled resultChecking, as it calls twice traceEndTx

Signed-off-by: Francois Bojarski <francois.bojarski@consensys.net>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 1006-range-6032696-6032699

OlivierBBB committed:
fix(romlex): fixed deployment information used in creation of ROM chunk

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -103,7 +104,7 @@ public int getCodeFragmentIndexByMetadata(final ContractMetadata metadata) {
         String.format(
             "RomChunk with address %s, deployment number %s and deploymentStatus %s not found",
-            metadata.address(), metadata.deploymentNumber(), !metadata.underDeployment()));
+            metadata.address(), metadata.deploymentNumber(), metadata.underDeployment()));
```
This was printing the wrong information in the error message.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/romlex/RomLex.java:
```diff
@@ -188,9 +189,8 @@ public void callRomLex(final MessageFrame frame) {
-        final int depNumber = hub.deploymentNumberOf(contractAddress);
-        final boolean depStatus = hub.deploymentStatusOf(contractAddress);
+        final int depNumber = hub.deploymentNumberOf(calledAddress);
+        final boolean depStatus = hub.deploymentStatusOf(calledAddress);
```
The `frame.contractAddress` is the address of the bytecode executing in the current execution context (or frame.) What we need is the deployment information of the code we are trying to `EXTCODECOPY` from

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/copy/ExtCodeCopySection.java:
```diff
@@ -122,8 +122,10 @@ public ExtCodeCopySection(Hub hub) {
     accountDoingFragment.requiresRomlex(triggerCfi);
+    if (triggerCfi) {
+      hub.romLex().callRomLex(hub.messageFrame());
+    }
```
We should trigger the ROM and ROM_LEX modules iff it is deemed necessary (`trigger_ROMLEX = true` or in the old notation `triggerCfi = true`)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/halt/StopSection.java:
```diff
@@ -50,7 +50,7 @@ public StopSection(Hub hub) {
 
     hubStamp = hub.stamp();
-    address = hub.currentFrame().accountAddress();
+    address = hub.messageFrame().getContractAddress();
```
We care about the address of the code that's currently running, not about the address of the account that's running it.

OlivierBBB committed:
re-enabled (now passing) tests

- TwoPlusTwoTest
- OobRdcTest

OlivierBBB committed:
spotless

OlivierBBB committed:
fix(storageSlotIdentifier): fixed (?) issue with deployment number discrepancy

OlivierBBB committed:
spotless

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/storage/StorageFragment.java:
```diff
@@ -43,20 +43,35 @@ public final class StorageFragment implements TraceFragment {
+    int x = 42;
+
+    checkArgument(match);
+    checkArgument(containsActualStorageSlotIdentifier);
```
TODO: remove, this was for testing purposes

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/storage/StorageFragment.java:
```diff
@@ -43,20 +43,35 @@ public final class StorageFragment implements TraceFragment {
+    int x = 42;
+
+    checkArgument(match);
+    checkArgument(containsActualStorageSlotIdentifier);
```
TODO: remove, this was for testing purposes

letypequividelespoubelles replied:
still here ?

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 1006-range-6032696-6032699

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/storage/StorageFragmentPurpose.java:
```diff
@@ -12,10 +12,14 @@
+    SSTORE_UNDOING,
+
+    PRE_WARMING
+};
```
Mostly for debugging, though we may want to do this more generally: have a system whereby we can tag / label rows (i.e. fragments) with auxiliary information such as
- hub stamp (if not accessible other wise)
- instruction
- "purpose" of the row

The `purpose` business will be useful for debugging CALL's etc ... anything with many rows with complex purposes.

Another tagging we should have present is to tag a `CallFrame` according to whether it was created by
- `CALL` / `STATICCALL`
- `DELEGATECALL` / `CALLCODE`
- `CREATE` / `CREATE2`

The distinction MESSAGE_CALL / CONTRACT_CREATION is too coarse, that is: we must refine the MESSAGE_CALL category

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/SloadSection.java:
```diff
@@ -43,23 +46,23 @@ public class SloadSection extends TraceSection implements PostRollbackDefer {
+    valueOriginal =
+        EWord.of(worldView.get(address).getOriginalStorageValue(UInt256.fromBytes(storageKey)));
+    valueCurrent = EWord.of(worldView.get(address).getStorageValue(UInt256.fromBytes(storageKey)));
+    exceptions = hub.pch().exceptions();
```
The disparity between the underlying addresses (accountAddress vs bytecodeAddress) seems to have caused the above.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/SstoreSection.java:
```diff
@@ -46,21 +49,21 @@ public class SstoreSection extends TraceSection implements PostRollbackDefer {
+    valueOriginal =
+        EWord.of(worldView.get(address).getOriginalStorageValue(UInt256.fromBytes(storageKey)));
+    valueCurrent = EWord.of(worldView.get(address).getStorageValue(UInt256.fromBytes(storageKey)));
+    valueNext = EWord.of(hub.messageFrame().getStackItem(1));
```
Same remark about the `address of the underlying account` vs `address whose code is currently executing`.

OlivierBBB committed:
better names + quality of life methods in the HUB

OlivierBBB committed:
spotless

DavePearce committed:
Use `ReferenceTestWorldState` instead of `ToyWorld`

This puts in place a temporary fix which converts `ToyWorld` over into
`ReferenceTestWorldState` to use when executing transactions.  The
conversion could be done more elegantly, but the goal here is to get
something out for people to use ASAP.  Eventually, we would then remove
`ToyWorld` altogether.

DavePearce committed:
Updating replay file for 6020023-6020029

OlivierBBB committed:
fix: don't check agreement on final state for now ...

OlivierBBB committed:
Merge branch 'arith-dev' into 1006-range-6032696-6032699

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/AccountSnapshot.java:
```diff
@@ -42,56 +43,61 @@ public class AccountSnapshot {
-  public AccountSnapshot incrementBalance(Wei quantity) {
-    this.balance = this.balance.add(quantity);
-    return this;
+  public static AccountSnapshot canonical(
```
What's the purpose of this ? We already have a `canonical` that does the same, wo the world in the signature, using `hub.messageFrame().getWorldUpdater()`

letypequividelespoubelles replied:
oh my bad, I remember we use it in startTx where we don't have the world through the frame

letypequividelespoubelles replied:
anyway, we should merge the two method

OlivierBBB replied:
> I remember we use it in startTx where we don't have the world through the frame

It's for that kind of reason. We don't have access to the `MessageFrame` at certain points in time but we do have access to some `WorldView` or so.

letypequividelespoubelles replied:
Still they should both call the same method with account hub and address to not duplicate the underlying constructor with if null etc

OlivierBBB replied:
Yeah I agree.

OlivierBBB replied:
Done

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/AccountSnapshot.java:
```diff
@@ -128,87 +134,101 @@ public static AccountSnapshot fromAccount(
+        "Insufficient balance\n     Address: %s\n     Balance: %s\n     Value: %s"
+            .formatted(this.address, this.balance, quantity));
+
+    this.balance = this.balance.subtract(quantity);
```
so we keep `this` ? ;)

OlivierBBB replied:
Done

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -484,6 +485,8 @@ public void traceStartTransaction(final WorldView world, final Transaction tx) {
     state.enter();
     txStack.enterTransaction(world, tx, transients.block());
 
+    defers.scheduleForPostTransaction(txStack.current());
```
what's the purpose of this ?

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java:
```diff
@@ -214,6 +214,7 @@ public CreateSection(Hub hub) {
             this, hub.currentFrame()); // To get the success bit of the CREATE(2)
 
     hub.romLex().callRomLex(frame);
+    // hub.transients().conflation().deploymentInfo().newDeploymentAt(createeAddress);
```
?

OlivierBBB replied:
I took it out here as it's already done at `traceContextEnter` in `Hub.java` so this would be doing it twice

![image](https://github.com/user-attachments/assets/62c3783f-a28c-403d-87a0-b92eb0979514)

letypequividelespoubelles replied:
OK maybe add a dic comment to explain it ?

OlivierBBB replied:
I removed it altogether (the comment.)

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxFinalizationSection.java:
```diff
@@ -30,202 +33,250 @@
+    recipientSnapshotBeforeFinalization = AccountSnapshot.canonical(hub, recipientAddress);
+    coinbaseSnapshotBeforeTxFinalization = AccountSnapshot.canonical(hub, coinbaseAddress);
+
+    //  TODO: re-enable checks
```
are the checks creatuing java exfeption ?

OlivierBBB replied:
Yes, they were last time I checked

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxSkippedSection.java:
```diff
@@ -46,93 +47,106 @@ public TxSkippedSection(
-    // Miner
-    this.addFragment(
+    // coinbase account fragment
+    AccountFragment coinbaseAccountFragment =
```
all fragments could be final

OlivierBBB replied:
Done

letypequividelespoubelles reviewed (approved):
LGTM

OlivierBBB committed:
fix(skip): made various fragments final

OlivierBBB committed:
fix(this): fixed this :D

OlivierBBB committed:
ras

OlivierBBB committed:
spotless

OlivierBBB committed:
fix(AccountSnapshot): unified the "canonical" methods

OlivierBBB committed:
ras
